sinabs/backend/dynapcnn/dynapcnnnetwork_module.py

```diff
@@ -1,99 +1,253 @@
+        sources = dict()
+
+        for src, trg in dcnnl_edges:
+            if trg in sources:
+                sources[trg].append(src)
+            else:
+                sources[trg] = [src]
+
+        return sources
 
-    def _spot_merging_points(
-        self, dcnnl_edges: list
+    # TODO: Probably not needed
+    def get_merging_points(
+        self, node_source_map: Dict[int, List[int]]
     ) -> Dict[int, Dict[Tuple, sl.Merge]]:
-        """Loops throught the edges of the computational graph from a `DynapcnnNetwork` to flag with nodes need
-        input from a `Merge` layer and what the arguments of this layer should be.
+        """Find nodes within `dcnnl_edges` that have multiple sources.
 
         Parameters
         ----------
-        - dcnnl_edges (list): tuples representing the output->input mapping between `DynapcnnLayer` instances
-            that have been used as configuration for each core `CNNLayerConifg`.
+        - node_source_map: Dict that maps to each layer index (int) a list of
+            indices of layers that act as input source to this node
+
+        Returns
+        -------
+        - Dict that for each layer with more than one input source maps its index
+            (int) to a nested dict with two entries:
+                * "sources": Set of indices of all source layers to this layer
+                * "merge_layer": A `Merge` layer instance
         """
+        return {
+            tgt: {"sources": sources, "merge_layer": sl.Merge()}
+            for tgt, sources in node_source_map
+            if len(sources) > 1
+        }
 
-        nodes_with_merge_input = {}
+    def forward(
+        self, x, return_complete: bool = False
+    ) -> Union[Tensor, Dict[int, Dict[int, Tensor]]]:
+        """Perform a forward pass through all dynapcnn layers
+        The `setup_dynapcnnlayer_graph` method has to be executed beforehand.
 
-        for edge in dcnnl_edges:
-            trg_node = edge[1]
-            fan_in = 0
-            src_nodes = []
-
-            # counts the fan-in for each target node `trg_node`.
-            for edge_inner in dcnnl_edges:
-                if edge_inner[1] == trg_node:
-                    # fan-in update.
-                    fan_in += 1
-                    src_nodes.append(edge_inner[0])
-
-            if fan_in == 2 and trg_node not in nodes_with_merge_input:
-                # node needs input from a `Merge` layer: instantiate `Merge` and its arguments.
-                nodes_with_merge_input[trg_node] = {
-                    "sources": tuple(src_nodes),
-                    "merge": sl.Merge(),
+        Parameters
+        ----------
+        x: Tensor that serves as input to network. Is passed to all layers
+            that are marked as entry points
+        return_complete: bool that indicates whether all layer outputs should
+            be return or only those with no further destinations (default)
+
+        Returns
```

Comment:
The type of return values now depend on multiple factors. I guess we should discuss if this makes sense.

Reply:
Another approach could be to not only mark "entry points" but also "exit points", if we can get those from the NIR/torch graph.

Reply:
.. I don't think we can get them from the torch graph in a straightforward fashion. We could get them the same way we get the entry nodes, simply by defining them as nodes with no target. However, that doesn't really give us anything beyond what we are doing. I guess for now we can leave it as it is.